### PR TITLE
8290700: Optimize AArch64 nmethod entry barriers

### PR DESCRIPTION
**Pull request by fisk**
The original nmethod entry barrier supported only concurrent patching of data and was used by ZGC to solve concurrent class unloading problems. Now it is starting to see more uses. Notably, loom uses nmethod entry barriers to figure out what nmethods have been seen on-stack, needed to remove nmethods safely. However, the concurrent data patching variation was too slow for loom (showed a few regressions), so I brought over a faster nmethod entry barrier that we use in the generational ZGC repo, which additionally handles concurrent patching of data and instructions, which is needed there. 
However, I still see some small regressions. So for the uses in loom, the classic GCs don't really patch anything interesting concurrently. This leads to the following possible enhancements to improve the situation: 

1. Make a dedicated nmethod entry barrier for GCs that don't patch data nor code concurrently, consisting of basically only a conditional branch. There is no need to penalize STW GCs with seat belts protecting against concurrent races that simply do not exist.

2. Move the "guard" word and call into the VM trampoline, to an out-of-line stub towards the end of the nmethod, ensuring instruction caches are not polluted by non-hot instructions at the nmethod entry. Some machines also better optimize the branch-not-taken path of conditional branches.

With these optimizations, the small regressions did go away.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290700](https://bugs.openjdk.org/browse/JDK-8290700): Optimize AArch64 nmethod entry barriers

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9574/head:pull/9574` \
`$ git checkout pull/9574`

Update a local copy of the PR: \
`$ git checkout pull/9574` \
`$ git pull https://git.openjdk.org/jdk pull/9574/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9574`

View PR using the GUI difftool: \
`$ git pr show -t 9574`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9574.diff">https://git.openjdk.org/jdk/pull/9574.diff</a>

</details>

### CODE REVIEW

**Comment by fisk** [comment]
Note that this PR is dependent on https://github.com/openjdk/jdk/pull/9569/

**Review by vnkozlov — APPROVED** [review]
Looks reasonable to me.
Someone familiar with aarch64 code have to review it.

**Comment by fisk** [comment]
Thanks for the review, @vnkozlov!

File: src/hotspot/cpu/aarch64/gc/shared/barrierSetNMethod_aarch64.cpp [hunk 1]
```diff
@@ -37,29 +37,62 @@
...
+      // We find it using the RelocIterator.
+      RelocIterator iter(nm);
+      while (iter.next()) {
+        if (iter.type() == relocInfo::entry_guard_type) {
+          entry_guard_Relocation* const reloc = iter.entry_guard_reloc();
+          return reinterpret_cast<int*>(reloc->addr());
+        }
+      }
+      ShouldNotReachHere();
+    }
```
Comment (dean-long) [inline src/hotspot/cpu/aarch64/gc/shared/barrierSetNMethod_aarch64.cpp:85]: I'm guessing this function is not performance-critical.  If it was, then we could consider adding a new field to the nmethod to keep track of the offset for the guard.

**Review by dean-long — APPROVED** [review]

File: src/hotspot/cpu/aarch64/gc/shared/barrierSetNMethod_aarch64.cpp (hunk 1 above)
Comment (fisk) [reply src/hotspot/cpu/aarch64/gc/shared/barrierSetNMethod_aarch64.cpp:85]: Yeah I thought about adding a new entry to CodeOffsets, but it seemed like this C2-only AArch64-only detail seemed a bit noisy in the shared code then. And it didn't seem warranted, given how cold this path is by design.

**Comment by fisk** [comment]
Thanks for the review @dean-long!

**Pull request closed without merging**